monitoring/exporter/stackdriver/project_data.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stackdriver
+
+import (
+	"fmt"
+
+	"go.opencensus.io/tag"
+	"google.golang.org/api/support/bundler"
+	metricpb "google.golang.org/genproto/googleapis/api/metric"
+	mpb "google.golang.org/genproto/googleapis/monitoring/v3"
+)
+
+// MaxTimeSeriePerUpload is the maximum number of time series that stackdriver accepts. Only test
+// may change this value.
+var MaxTimeSeriesPerUpload = 200
+
+// projectData contain per-project data in exporter. It should be created by newProjectData()
+type projectData struct {
+	parent    *Exporter
+	projectID string
+	// We make bundler for each project because call to monitoring RPC can be grouped only in
+	// project level
+	bndler *bundler.Bundler
+}
+
+func (e *Exporter) newProjectData(projectID string) *projectData {
+	pd := &projectData{
+		parent:    e,
+		projectID: projectID,
+	}
+
+	pd.bndler = newBundler((*RowData)(nil), pd.uploadRowData)
+	// Set options for bundler if they are provided by users.
+	if 0 < e.opts.BundleDelayThreshold {
+		pd.bndler.DelayThreshold = e.opts.BundleDelayThreshold
+	}
+	if 0 < e.opts.BundleCountThreshold {
+		pd.bndler.BundleCountThreshold = e.opts.BundleCountThreshold
+	}
+	return pd
+}
+
+// uploadRowData is called by bundler to upload row data, and report any error happened meanwhile.
+func (pd *projectData) uploadRowData(bundle interface{}) {
+	exp := pd.parent
+	rds := bundle.([]*RowData)
+
+	// reqRds contains RowData objects those are uploaded to stackdriver at given iteration.
+	// It's main usage is for error reporting. For actual uploading operation, we use req.
+	// remainingRds are RowData that has not been processed at all.
+	var reqRds, remainingRds []*RowData
```

Comment:
This whole flow seems a little hard to read for me. How about restructuring this code as follows:

func uploadRowData(bundle) {
  - get the slice of RowData from the bundle
  - iterate through the slice, and for each RowData {
    - call a func named makeTS() which returns one &mpb.TimeSeries
    - maintain a slice of TimeSeries objects here, and append the one returned in the previous line to that slice
   - if the slice has reached the required length, call another function (or just inline it here) to actually make the RPC
  }
  - if you have items still in the timeSeries slice, make another call to send an RPC
}

What do you think?

Reply:
I generally followed your suggestion.